src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs

```diff
@@ -370,126 +418,410 @@ private sealed class SendFileOperation : WriteOperation
+            // to try to perform the enqueued operations.
+            // If an operation is successfully performed, we remove it from the queue,
+            // enqueue another threadpool workitem to process the next item in the queue (if any),
+            // and call the user's completion callback.
+            // If we successfully process all enqueued operations, then the state becomes Ready;
+            // otherwise, the state becomes Waiting and we wait for another epoll notification.
+
+            private enum QueueState
+            {
+                Ready = 0,          // Indicates that data MAY be available on the socket.
+                                    // Queue must be empty.
+                Waiting = 1,        // Indicates that data is definitely not available on the socket.
+                                    // Queue must not be empty.
+                Processing = 2,     // Indicates that a thread pool item has been scheduled (and may 
+                                    // be executing) to process the IO operations in the queue.
+                                    // Queue must not be empty.
+                Stopped = 3,        // Indicates that the queue has been stopped because the 
+                                    // socket has been closed.
+                                    // Queue must be empty.
+            }
+
+            // These fields define the queue state.
+
+            private QueueState _state;      // See above
+            private int _sequenceNumber;    // This sequence number is updated when we receive an epoll notification.
+                                            // It allows us to detect when a new epoll notification has arrived
+                                            // since the last time we checked the state of the queue.
+                                            // If this happens, we MUST retry the operation, otherwise we risk
+                                            // "losing" the notification and causing the operation to pend indefinitely.
+            private AsyncOperation _tail;   // Queue of pending IO operations to process when data becomes available.
+
+            // The _queueLock is used to ensure atomic access to the queue state above.
+            // The lock is only ever held briefly, to read and/or update queue state, and
+            // never around any external call, e.g. OS call or user code invocation.
             private object _queueLock;
-            private AsyncOperation _tail;
 
-            public QueueState State { get; set; }
-            public bool IsStopped { get { return State == QueueState.Stopped; } }
-            public bool IsEmpty { get { return _tail == null; } }
-            public object QueueLock { get { return _queueLock; } }
+            private LockToken Lock() => new LockToken(_queueLock);
+
+            private static WaitCallback s_processingCallback =
```

Comment:
It's initialized as part of being declared.  You're saying it's also assigned to elsewhere?  Where?